File: solutions/java/ricdotnet/src/main/java/com/google/VideoPlayer.java
package com.google;

import java.util.*;

public class VideoPlayer {

  Scanner scanner;

  private final VideoLibrary videoLibrary;
  private Video video;
  private Video playing;
  private Boolean isPaused = false;

  private final List<VideoPlaylist> playlists = new ArrayList<>();
  private List<Video> searchResult; // cannot be final and I need it to always be empty from the start

  public VideoPlayer() {
    this.videoLibrary = new VideoLibrary();
  }

  public void numberOfVideos() {
    System.out.printf("%s videos in the library%n", videoLibrary.getVideos().size());
  }

  /**
   * DONE
   */
  public void showAllVideos() {
//    System.out.println("showAllVideos needs implementation");
    System.out.println("Here's a list of all available videos:");

    List<Video> tempList = videoLibrary.getVideos();
    tempList.sort(Comparator.naturalOrder());
    for (Video video : tempList) {
      System.out.printf("%s (%s) %s %s\n", video.getTitle(), video.getVideoId(),
          video.getTags().toString().replaceAll(",", ""), video.getIsFlagged() ? "- FLAGGED " +
              "(reason: " + flagReasonHelper(video.getVideoId()) + ")" : "");
    }
  }

  /**
   * DONE
   *
   * @param videoId
   */
  public void playVideo(String videoId) {
//    System.out.println("playVideo needs implementation");
    if (getVideo(videoId) == null) {
      System.out.println("Cannot play video: Video does not exist.");
      return;
    }

    if (getVideo(videoId).getIsFlagged()) {
      System.out.printf("Cannot play video: Video is currently flagged (reason: %s)\n",
          getVideo(videoId).getFlaggedReason() == null ? "Not supplied" : getVideo(videoId).getFlaggedReason());
      return;
    }

    if (isPaused)
      isPaused = false; //if the video exists then it will set isPaused to false as a new video will start playing
    if (playing != null) {
      System.out.printf("Stopping video: %s\n", playing.getTitle());
    }
    playing = getVideo(videoId);
    System.out.printf("Playing video: %s\n", playing.getTitle());
  }

  /**
   * DONE
   */
  public void stopVideo() {
//    System.out.println("stopVideo needs implementation");
    if (playing != null) {
      System.out.printf("Stopping video: %s\n", playing.getTitle());
      playing = null;
      return;
    }

    System.out.println("Cannot stop video: No video is currently playing");
  }

  /**
   * DONE
   */
  public void playRandomVideo() {
//    System.out.println("playRandomVideo needs implementation");

    if (isAllFlaggedHelper()) {
      System.out.println("No videos available");
      return;
    }

    video = videoLibrary.getVideos().get((int)
        Math.floor(Math.random() * videoLibrary.getVideos().size()));


//    if(video.getIsFlagged()) {
//      System.out.printf("Cannot play video: Video is currently flagged (reason: %s)",
//          video.getFlaggedReason() == null ? "Not supplied" : video.getFlaggedReason());
//      return;
//    }

    playVideo(video.getVideoId());
  }

  /**
   * DONE
   */
  public void pauseVideo() {
//    System.out.println("pauseVideo needs implementation");
    if (playing == null) {
      System.out.println("Cannot pause video: No video is currently playing.");
      return;
    }

    if (!isPaused) {
      System.out.printf("Pausing video: %s\n", playing.getTitle());
      isPaused = true;
      return;
    }

    System.out.printf("Video already paused: %s\n", playing.getTitle());
  }

  /**
   * DONE
   */
  public void continueVideo() {
//    System.out.println("continueVideo needs implementation");
    if (playing == null) {
      System.out.println("Cannot continue video: No video is currently playing");
      return;
    }

    if (!isPaused) {
      System.out.println("Cannot continue video: Video is not paused");
      return;
    }

    isPaused = false;
    System.out.printf("Continuing video: %s", playing.getTitle());
  }

  /**
   * DONE
   */
  public void showPlaying() {
//    System.out.println("showPlaying needs implementation");
    if (playing == null) {
      System.out.println("No video is currently playing");
      return;
    }

    System.out.printf("Currently playing: %s (%s) %s %s\n",
        playing.getTitle(), playing.getVideoId(),
        playing.getTags().toString().replaceAll(",", ""),
        (isPaused) ? "- PAUSED" : "");
  }

  /**
   * DONE
   *
   * @param playlistName
   */
  public void createPlaylist(String playlistName) {
//    System.out.println("createPlaylist needs implementation");
    //check for repeated playlists
    if (getPlaylist(playlistName) != null) {
      System.out.println("Cannot create playlist: A playlist with the same name already exists");
      return;
    }

    playlists.add(new VideoPlaylist(playlistName));
    System.out.printf("Successfully created new playlist: %s\n", playlistName);
  }

  /**
   * DONE
   *
   * @param playlistName
   * @param videoId
   */
  public void addVideoToPlaylist(String playlistName, String videoId) {
//    System.out.println("addVideoToPlaylist needs implementation");
//    playlist = getPlaylist(playlistName);
    if (getPlaylist(playlistName) == null) {
      System.out.printf("Cannot add video to %s: Playlist does not exist\n", playlistName);
      return;
    }

    if (getVideo(videoId) == null) {
      System.out.printf("Cannot add video to %s: Video does not exist\n", playlistName);
      return;
    }

    if (getVideo(videoId).getIsFlagged()) {
      System.out.printf("Cannot add video to %s: Video is currently flagged (reason: %s)\n",
          playlistName,
          getVideo(videoId).getFlaggedReason() == null ? "Not supplied" : getVideo(videoId).getFlaggedReason());
      return;
    }

    if (getPlaylist(playlistName).getVideos().contains(getVideo(videoId))) {
      System.out.printf("Cannot add video to %s: Video already added\n", playlistName);
      return;
    }

    getPlaylist(playlistName).getVideos().add(getVideo(videoId));
    System.out.printf("Added video to %s: %s\n", playlistName, getVideo(videoId).getTitle());

  }

  /**
   * DONE
   */
  public void showAllPlaylists() {
//    System.out.println("showAllPlaylists needs implementation");
    if (playlists.size() == 0) {
      System.out.println("No playlists exist yet");
      return;
    }

    System.out.println("Showing all playlists:");
    List<VideoPlaylist> tempList = playlists;
    tempList.sort(Comparator.naturalOrder());
    for (VideoPlaylist videoPlaylist : playlists) {
      System.out.println(videoPlaylist.getName());
    }
  }

  /**
   * DONE
   *
   * @param playlistName
   */
  public void showPlaylist(String playlistName) {
//    System.out.println("showPlaylist needs implementation");
    if (getPlaylist(playlistName) == null) {
      System.out.printf("Cannot show playlist %s: Playlist does not exist\n", playlistName);
      return;
    }

    System.out.printf("Showing playlist: %s (%s videos)\n", playlistName,
        getPlaylist(playlistName).getVideos().size());
    if (getPlaylist(playlistName).getVideos().size() == 0) {
      System.out.println("No videos here yet");
      return;
    }

    for (Video video : getPlaylist(playlistName).getVideos()) {
      System.out.printf("%s (%s) %s %s\n", video.getTitle(), video.getVideoId(),
          video.getTags().toString().replaceAll(",", ""), video.getIsFlagged() ? "- FLAGGED " +
              "(reason: " + flagReasonHelper(video.getVideoId()) + ")" : "");
    }
  }

  /**
   * DONE
   *
   * @param playlistName
   * @param videoId
   */
  public void removeFromPlaylist(String playlistName, String videoId) {
//    System.out.println("removeFromPlaylist needs implementation");
    if (getPlaylist(playlistName) == null) {
      System.out.printf("Cannot remove video from %s: Playlist does not exist\n", playlistName);
      return;
    }

    if (getVideo(videoId) == null) {
      System.out.printf("Cannot remove video from %s: Video does not exist\n", playlistName);
      return;
    }

    if (!getPlaylist(playlistName).getVideos().contains(getVideo(videoId))) {
      System.out.printf("Cannot remove video from %s: Video is not in playlist\n", playlistName);
      return;
    }

    getPlaylist(playlistName).getVideos().remove(getVideo(videoId));
    System.out.printf("Removed video from %s: %s\n", playlistName, getVideo(videoId).getTitle());

  }

  /**
   * DONE
   *
   * @param playlistName
   */
  public void clearPlaylist(String playlistName) {
//    System.out.println("clearPlaylist needs implementation");
    if (getPlaylist(playlistName) == null) {
      System.out.printf("Cannot clear playlist %s: Playlist does not exist\n", playlistName);
      return;
    }

    getPlaylist(playlistName).getVideos().clear();
    System.out.printf("Successfully removed all videos from %s\n", playlistName);
  }

  /**
   * DONE
   *
   * @param playlistName
   */
  public void deletePlaylist(String playlistName) {
//    System.out.println("deletePlaylist needs implementation");
    if (getPlaylist(playlistName) == null) {
      System.out.printf("Cannot delete playlist %s: Playlist does not exist\n", playlistName);
      return;
    }

    playlists.remove(getPlaylist(playlistName));
    System.out.printf("Deleted playlist: %s\n", playlistName);
  }

  /**
   * DONE
   *
   * @param searchTerm
   */
  public void searchVideos(String searchTerm) {
//    System.out.println("searchVideos needs implementation");

    scanner = new Scanner(System.in);
    searchResult = new ArrayList<>();

    for (Video video : videoLibrary.getVideos()) {
      if (video.getTitle().toLowerCase().contains(searchTerm.toLowerCase()) && !video.getIsFlagged()) {
        searchResult.add(video);
      }
    }

    if (searchResult.size() == 0) {
      System.out.printf("No search results for %s", searchTerm);
      return;
    }

    printSearchResult(searchTerm, searchResult);

    System.out.println("Would you like to play any of the above? If yes, specify the number of the video.\n" +
        "If your answer is not a valid number, we will assume it's a no.");

    String input = scanner.next();

    try {
      int option = Integer.parseInt(input);
      if (option > 0 && option <= searchResult.size()) {
        playVideo(searchResult.get(option - 1).getVideoId());
      }
    } catch (NumberFormatException ignore) {
    }

  }

  /**
   * DONE
   *
   * @param videoTag
   */
  public void searchVideosWithTag(String videoTag) {
//    System.out.println("searchVideosWithTag needs implementation");
    scanner = new Scanner(System.in);
    searchResult = new ArrayList<>();

    if (videoTag.startsWith("#")) {
      for (Video video : videoLibrary.getVideos()) {
        for (String tag : videoLibrary.getVideo(video.getVideoId()).getTags()) {
          if (tag.equalsIgnoreCase(videoTag) && !video.getIsFlagged()) {
            searchResult.add(video);
          }
        }
      }
    }

    if (searchResult.size() == 0) {
      System.out.printf("No search results for %s", videoTag);
      return;
    }

    printSearchResult(videoTag, searchResult);

    System.out.println("Would you like to play any of the above? If yes, specify the number of the video.\n" +
        "If your answer is not a valid number, we will assume it's a no.");

    String input = scanner.next();

    try {
      int option = Integer.parseInt(input);
      if (option > 0 && option <= searchResult.size()) {
        playVideo(searchResult.get(option - 1).getVideoId());
      }
    } catch (NumberFormatException ignore) {
    }

  }

  /**
   * DONE
   *
   * @param videoId
   */
  public void flagVideo(String videoId) {
//        System.out.println("flagVideo needs implementation");

    // a return different than 0 will print an error message set on the helper method
    if (flagHelper(videoId) != 0) {
      return;
    }

    if (playing != null && playing.getVideoId().equals(videoId)) {
      stopVideo();
    }

    getVideo(videoId).setIsFlagged(true);
    System.out.printf("Successfully flagged video: %s (reason: Not supplied)\n",
        getVideo(videoId).getTitle());
  }

  /**
   * DONE
   *
   * @param videoId
   * @param reason
   */
  public void flagVideo(String videoId, String reason) {
//        System.out.println("flagVideo needs implementation");
    if (flagHelper(videoId) != 0) {
      return;
    }

    if (playing != null && playing.getVideoId().equals(videoId)) {
      stopVideo();
    }

    getVideo(videoId).setIsFlagged(true);
    getVideo(videoId).setFlaggedReason(reason);

    System.out.printf("Successfully flagged video: %s (reason: %s)\n",
        getVideo(videoId).getTitle(), getVideo(videoId).getFlaggedReason());
  }

  /**
   * DONE
   *
   * @param videoId
   */
  public void allowVideo(String videoId) {
//        System.out.println("allowVideo needs implementation");
    if (getVideo(videoId) == null) {
      System.out.println("Cannot remove flag from video: Video does not exist");
      return;
    }

    if (!getVideo(videoId).getIsFlagged()) {
      System.out.println("Cannot remove flag from video: Video is not flagged");
      return;
    }

    getVideo(videoId).setIsFlagged(false);
    getVideo(videoId).setFlaggedReason(null);

    System.out.printf("Successfully removed flag from video: %s\n", getVideo(videoId).getTitle());
  }


  /**
   * HELPER METHODS
   */
  private Video getVideo(String videoId) {
    for (Video video : videoLibrary.getVideos()) {
      if (video.getVideoId().equals(videoId)) {
        return video;
      }
    }
    return null;
  }

  private VideoPlaylist getPlaylist(String playlistName) {
    for (VideoPlaylist videoPlaylist : playlists) {
      if (playlistName.equalsIgnoreCase(videoPlaylist.getName())) {
        return videoPlaylist;
      }
    }
    return null;
  }

  private void printSearchResult(String term, List<Video> list) {
    searchResult.sort(Comparator.naturalOrder()); //sort the list
    System.out.printf("Here are the results for %s:\n", term);
    for (Video video : list) {
      System.out.printf("%s) %s (%s) %s\n",
          searchResult.indexOf(video) + 1, video.getTitle(), video.getVideoId(),
          video.getTags().toString().replaceAll(",", ""));
    }
  }

  private int flagHelper(String videoId) {
    if (getVideo(videoId) == null) {
      System.out.println("Cannot flag video: Video does not exist");
      return 1;
    }

    if (getVideo(videoId).getIsFlagged()) {
      System.out.println("Cannot flag video: Video is already flagged");
      return 1;
    }

    return 0;
  }

  private Boolean isAllFlaggedHelper() {
    int count = 0;
    for (Video video : videoLibrary.getVideos()) {
      if (!video.getIsFlagged())
        count++;
    }
    return count == 0;
  }

  private String flagReasonHelper(String videoId) {
    if (getVideo(videoId).getFlaggedReason() != null) {
      return getVideo(videoId).getFlaggedReason();
    }

    return "no reason";
  }

  private int sortByTitle(Video v1, Video v2) {
    return v1.getTitle().compareTo(v2.getTitle());
  }
}

File: solutions/python/solution3/src/video_player.py
"""A video player class."""
import random

from .video_library import VideoLibrary
from .video_playlist import PlaylistLibrary

class PlayState:
    def __init__(self, v=None):
        self.video = v
        self.paused = False

class VideoPlayer:
    """A class used to represent a Video Player."""

    def __init__(self):
        """The VideoPlayer class is initialized."""
        self._video_library = VideoLibrary()
        self._current_playing = None
        self._playlists = PlaylistLibrary()

    def number_of_videos(self):
        num_videos = len(self._video_library.get_all_videos())
        print(f"{num_videos} videos in the library")

    def show_all_videos(self):
        """Returns all videos."""
        print("Here's a list of all available videos:")
        for v in sorted(self._video_library.get_all_videos(), key=lambda v: v.title):
            print(f"  {v.format()}")

    def play_video(self, video_id):
        """Plays the respective video.

        Args:
            video_id: The video_id to be played.
        """
        video = self._video_library.get_video(video_id)
        if video is None:
            print("Cannot play video: Video does not exist")
            return
        if video.flag:
            print(f"Cannot play video: Video is currently flagged (reason: {video.flag})")
            return
        self.stop_video(silent=True)
        print("Playing video:", video.title)
        self._current_playing = PlayState(video)

    def stop_video(self, silent=False, very_silent=False):
        """Stops the current video."""
        if self._current_playing is None:
            if not silent and not very_silent:
                print("Cannot stop video: No video is currently playing")
            return False
        if not very_silent:
            print("Stopping video:", self._current_playing.video.title)
        self._current_playing = None
        self._paused = False
        return True

    def play_random_video(self):
        """Plays a random video from the video library."""
        videos = self._video_library.get_all_unflagged_videos()
        if not videos:
            print("No videos available")
            return
        v = random.choice(videos)
        self.play_video(v.video_id)

    def pause_video(self):
        """Pauses the current video."""
        if self._current_playing is None:
            print("Cannot pause video: No video is currently playing")
            return
        if self._current_playing.paused:
            print("Video already paused:", self._current_playing.video.title)
            return
        self._current_playing.paused = True
        print("Pausing video:", self._current_playing.video.title)

    def continue_video(self):
        """Resumes playing the current video."""
        if self._current_playing is None:
            print("Cannot continue video: No video is currently playing")
            return
        if not self._current_playing.paused:
            print("Cannot continue video: Video is not paused")
            return
        print("Continuing video:", self._current_playing.video.title)
        self._current_playing.paused = False

    def show_playing(self):
        """Displays video currently playing."""
        if self._current_playing is None:
            print("No video is currently playing")
            return
        print("Currently playing:", self._current_playing.video.format(),
              "- PAUSED" * self._current_playing.paused)

    def create_playlist(self, playlist_name):
        """Creates a playlist with a given name.

        Args:
            playlist_name: The playlist name.
        """
        if not self._playlists.create_playlist(playlist_name):
            return
        print("Successfully created new playlist:", playlist_name)

    def add_to_playlist(self, playlist_name, video_id):
        """Adds a video to a playlist with a given name.

        Args:
            playlist_name: The playlist name.
            video_id: The video_id to be added.
        """
        p = self._playlists.get_playlist(playlist_name)
        if p is None:
            print(f"Cannot add video to {playlist_name}: Playlist does not exist")
            return
        v = self._video_library.get_video(video_id)
        if v is None:
            print(f"Cannot add video to {playlist_name}: Video does not exist")
            return
        if v.flag:
            print(f"Cannot add video to {playlist_name}: Video is currently flagged (reason: {v.flag})")
            return
        if not p.add_video(v):
            print(f"Cannot add video to {playlist_name}: Video already added")
            return
        print(f"Added video to {playlist_name}: {v.title}")
        

    def show_all_playlists(self):
        """Display all playlists."""

        print("show_all_playlists needs implementation")

    def show_playlist(self, playlist_name):
        """Display all videos in a playlist with a given name.

        Args:
            playlist_name: The playlist name.
        """
        print("show_playlist needs implementation")

    def remove_from_playlist(self, playlist_name, video_id):
        """Removes a video to a playlist with a given name.

        Args:
            playlist_name: The playlist name.
            video_id: The video_id to be removed.
        """
        p = self._playlists.get_playlist(playlist_name)
        if p is None:
            print(f"Cannot remove video from {playlist_name}: Playlist does not exist")
            return
        v = self._video_library.get_video(video_id)
        if v is None:
            print(f"Cannot remove video from {playlist_name}: Video does not exist")
            return
        if not p.remove_video(video_id):
            print(f"Cannot remove video from {playlist_name}: Video is not in playlist")
            return
        print(f"Removed video from {playlist_name}: {v.title}")


    def clear_playlist(self, playlist_name):
        """Removes all videos from a playlist with a given name.

        Args:
            playlist_name: The playlist name.
        """
        p = self._playlists.get_playlist(playlist_name)
        if p is None:
            print(f"Cannot clear playlist {playlist_name}: Playlist does not exist")
            return
        p.clear()
        print(f"Successfully removed all videos from {playlist_name}")

    def delete_playlist(self, playlist_name):
        """Deletes a playlist with a given name.

        Args:
            playlist_name: The playlist name.
        """
        self._playlists.delete_playlist(playlist_name)

    def show_playlist(self, playlist_name):
        """Display all videos in a playlist with a given name.

        Args:
            playlist_name: The playlist name.
        """
        p = self._playlists.get_playlist(playlist_name)
        if p is None:
            print(f"Cannot show playlist {playlist_name}: Playlist does not exist")
            return
        print("Showing playlist:", playlist_name)
        if not p.videos:
            print("  No videos here yet")
            return
        for v in p.videos:
            print(f"  {v.format()}")

    def show_all_playlists(self):
        """Display all playlists."""
        ps = self._playlists.get_all_playlists()
        if not ps:
            print("No playlists exist yet.")
            return
        print("Showing all playlists:")
        for p in sorted(ps, key=lambda p: p.playlist_id):
            print(f"  {p.playlist_id} ({len(p.videos)} videos)")

    def _search_selection(self, search_term, res):
        results = sorted(res, key=lambda v: v.title)
        if not results:
            print(f"No search results for {search_term}")
            return
        print(f"Here are the results for {search_term}:")
        for i, v in enumerate(results):
            print(f"  {i+1}) {v.format()}")
        print("Would you like to play any of the above? If yes, specify the number of the video.")
        print("If your answer is not a valid number, we will assume it's a no.")
        selection = input()
        if not selection.isnumeric():
            return
        sel = int(selection)
        if sel < 1 or sel > len(results):
            return
        self.play_video(results[sel - 1].video_id)

    def search_videos(self, search_term):
        """Display all the videos whose titles contain the search_term.

        Args:
            search_term: The query to be used in search.
        """
        self._search_selection(
            search_term,
            filter(
                lambda v: search_term.lower() in v.title.lower(),
                self._video_library.get_all_unflagged_videos()))

    def search_videos_tag(self, video_tag):
        """Display all videos whose tags contains the provided tag.

        Args:
            video_tag: The video tag to be used in search.
        """
        self._search_selection(
            video_tag,
            filter(
                lambda v: video_tag.lower() in {t.lower() for t in v.tags},
                self._video_library.get_all_unflagged_videos()))

    def flag_video(self, video_id, flag_reason="Not supplied"):
        """Mark a video as flagged.

        Args:
            video_id: The video_id to be flagged.
            flag_reason: Reason for flagging the video.
        """
        v = self._video_library.get_video(video_id)
        if v is None:
            print("Cannot flag video: Video does not exist")
            return
        if v.flag:
            print("Cannot flag video: Video is already flagged")
            return
        if self._current_playing is not None and self._current_playing.video.video_id == video_id:
            self.stop_video()
        v.set_flag(flag_reason)
        print(f"Successfully flagged video: {v.title} (reason: {flag_reason})")
        if self._current_playing and self._current_playing.video.video_id == video_id:
            self.stop_video(very_silent=True)

    def allow_video(self, video_id):
        """Removes a flag from a video.

        Args:
            video_id: The video_id to be allowed again.
        """
        v = self._video_library.get_video(video_id)
        if v is None:
            print("Cannot remove flag from video: Video does not exist")
            return
        if not v.flag:
            print("Cannot remove flag from video: Video is not flagged")
            return
        v.set_flag(flag=None)
        print("Successfully removed flag from video:", v.title)

